### PR TITLE
New Overrides (with tests)

### PR DESCRIPTION
**Pull request by rambleraptor**
These are the new overrides and their corresponding tests.

Additionally, the compiler will figure out which set of overrides are being used and apply the correct ones. Monitoring on TF now uses the new overrides and the others still use the old overrides.

-----------------------------------------------------------------
# [all]

### CODE REVIEW

**Comment by rambleraptor** [comment]
I'm expecting a lot of notes on this.

**Comment by danawillow** [comment]
This is a lot of code reviewers! Can you add a bit about who you're looking for for which parts?

**Comment by rambleraptor** [comment]
For sure:

@SirGitsalot - general code review
@TF folks - do you see any places this could terribly break you? My giant PR had me moving over all three providers with no changes + breaks.
@rileykarson - do you see a world in which this would help you change field nesting?

@all - the hope is that this change makes it easier to do what we want with the overrides (type changes were literally not possible in the last version!) and make it easier for us to make changes in the future. how can this code be changed to make those more accomplishable?

**Comment by rileykarson** [comment]
Agreed with @danawillow; imo there should be 1-2 primary reviewers and everyone else optional or for very specific parts.

Also, if we're able to do a live migration from old -> new overrides could we use the new ones somewhere in this PR? I am _really_ not a fan of changes that add code that doesn't do anything (but no-ops are great!) Resourcemanager is resource-light & has cross Ansible and Terraform support if it has to be an entire product at once.

**Comment by chrisst** [comment]
Following up on Riley's comment: you could open a separate PR based off this branch and using this branch as the base. It would allow somebody reviewing to see the new pattern in use and still allow you to keep the two conversations separate if that's what you're hoping for.

File: provider/overrides/resources.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 Google Inc.
+# Copyright 2017 Google Inc.
```
Comment (danawillow) [inline provider/overrides/resources.rb]: why this change?

**Comment by rambleraptor** [comment]
I figured out how to make the compiler accept either set of overrides. In the interim, we can move things over one-by-one.

There's a TF PR, but they seem unrelated.

File: provider/overrides/resources.rb (hunk 1 above)
Comment (rambleraptor) [reply provider/overrides/resources.rb]: good call. I've got the working version in another branch and it's been easier to copy-paste rather than git to get things updated. my bad!

**Comment by rileykarson** [comment]
Oh yup those Terraform PRs are just from bad magician timing

File: compiler.rb [hunk 2]
```diff
@@ -122,6 +122,13 @@
 
   provider_config = \
     Provider::Config.parse(provider_yaml_path, product_api, version)
+
+  # TODO(alexstephen): Remove when old overrides are deprecated.
+  if provider_config.is_a?(Array)
```
Comment (rileykarson) [inline compiler.rb]: This is confusing! `provider_config` can magically be one thing or another. Can we replace the first occurrence of `provider_config` with something like `product_config_or_product_config_and_api` and then explicitly choose one or the other?

File: provider/overrides/runner.rb [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    #       someProperty: !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'foobar' # replaces description
+    #       anotherProperty.someNestedProperty:
+    #         !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'baz'
+    #       anotherProperty[].someNestedPropertyInAnArray:
+    #         !ruby/object:Provider::MyProvider::PropertyOverride
+    #         description: 'baz'
+    #   ...
+    class Runner
```
Comment (rileykarson) [inline provider/overrides/runner.rb]: Classes called -er are often a code smell and not really necessary. We never use Runner without immediately calling `.build()` on it precisely once, so I'm not sure why we need a class here? This could probably just be a function.

File: provider/config.rb [hunk 4]
```diff
@@ -110,7 +123,8 @@ def validate
       overrides
 
       check_optional_property :files, Provider::Config::Files
-      check_property :overrides, Provider::ResourceOverrides
+      check_property :overrides, [Provider::ResourceOverrides,
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/config.rb]: It's an OR.

File: provider/overrides/runner.rb (hunk 3 above)
Comment (rambleraptor) [reply provider/overrides/runner.rb]: That's a valid point. You've seen me do this design pattern before.

Here's my rationale and I'm curious what others think:

I have this large chunk of code that's split up into many functions. I wanted to split this code off into some unit that's responsible for the overall action this code achieves. In addition, I want to be able to mark most of this code as not being necessary outside of some entry-level function.

In my mind, I've mostly described a class just with no instance variables.

File: compiler.rb [hunk 5]
```diff
@@ -120,8 +120,17 @@
     next
   end
 
-  provider_config = \
+  provider_config_or_array = \
     Provider::Config.parse(provider_yaml_path, product_api, version)
+
+  # TODO(alexstephen): Remove when old overrides are deprecated.
+  if provider_config_or_array.is_a?(Array)
+    product_api, provider_config = provider_config
```
Comment (rileykarson) [inline compiler.rb]: Also oops, sorry for the late addition here - can you explain with a comment why we're overriding the `product_api`?

Comment (rambleraptor) [reply compiler.rb]: Done! I think it'll make it easier to understand that the overrides are changing the entirety of the api object after calling `config.validate` too.

**Comment by rileykarson** [comment]
Just to double check, we're waiting on the second PR to be opened here before reviewing right?

**Comment by rambleraptor** [comment]
Please review this. This PR uses the new overrides on one product.

File: provider/overrides/terraform/property_override.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+module Provider
+  module Overrides
+    module Terraform
+      # Terraform-specific overrides to api.yaml.
+      class PropertyOverride < Provider::Overrides::PropertyOverride
+        # Collection of fields allowed in the PropertyOverride section for
+        # Terraform. All fields should be `attr_reader :<property>`
+        def self.attributes
+          Provider::Terraform::OverrideFields.attributes
```
Comment (rileykarson) [inline provider/overrides/terraform/property_override.rb]: I'm not sure I understand why we do this vs just including it inline. This looks like the only usage & we have to specify the fields in validate anyways?

File: provider/terraform/config.rb [hunk 7]
```diff
@@ -29,6 +29,14 @@ def resource_override
       def property_override
         Provider::Terraform::PropertyOverride
       end
+
+      def new_resource_override
```
Comment (rileykarson) [inline provider/terraform/config.rb]: This is called `new_` because these are the new overrides right? I would comment explaining that because the function name alone suggests that this will make a new resource override.

Also are there a lot of calls to `resource_override` or can we start using that name for the new ones now and call the old ones `old_resource_override`?

**Comment by slevenick** [comment]
Ok weird question. How does the new provider override functionality compare to merging two hashes? I ask because it seems to produce a pretty similar outcome to just loading two yaml files as hashes and calling merge on them.

**Comment by rambleraptor** [comment]
@slevenick that's an interesting idea. I'll do some experimenting.

**Comment by rileykarson** [comment]
I think we want more abstraction than merging 2 hashes so that we can make schema-modifying changes like convenience fields instead of just attribute-editing ones. Here's a few types of changes we need in the future as I defined them for Terraform:

> `convenience fields`: When we take a nested object field and top-level it.
> useful for nested objects with non-zero defaults.
> example: BackendService connection_draining.draining_timeout_sec -> connection_draining_timeout_sec
> 
> `split fields`: When we take a long string such as a new-style `name` self link and create multiple fields out of it.
> example: Redis name (projects/{projectId}/locations/{locationId}/instances/{instanceId}) ->  [name, project, region]
> 
> `derived fields`: When a field is based on the value of another field and is _not_ user specified.
> Some apis have fields that we always know the value of based on other fields such as new-style HealthCheck type fields. Some split fields require us to reconstruct information with a derived field such as Project having parent.[id, type] -> [folder_id, org_id]
> example: Project parent.[id, type] -> [folder_id, org_id]
> 
> `fine-grained resource`: When a nested field of a resource is sufficiently complex that a separate Terraform resource is required in order to express it.
> One common case is when a resource has an output-only field that's updated by an `add{{field}}` and `delete{{field}}` method on the resource.
> example: google_compute_network.peerings as google_compute_network_peering  (*Needs* to be fine grained as each network has to peer with the other, causing a cycle!)

File: provider/terraform/config.rb (hunk 7 above)
Comment (rambleraptor) [reply provider/terraform/config.rb]: I didn't want to touch the old overrides because this is already a long PR. The `new` part does refer to the new overrides. I'll add a comment.

File: provider/overrides/terraform/property_override.rb (hunk 6 above)
Comment (rambleraptor) [reply provider/overrides/terraform/property_override.rb]: We need a self.attributes to define the attributes that are accessible to the new overrides. The old way of doing this was some module magic that I still can't understand. Having a method makes it more explicit.

I did this OverrideFields thing so that I could share code between new overrides and old overrides.

File: provider/terraform/config.rb (hunk 7 above)
Comment (rileykarson) [reply provider/terraform/config.rb]: sgtm

**Comment by rambleraptor** [comment]
Since we're now building a brand new Api object, we can do more interesting things (like what Riley said above) and not have to worry about how past-state would interfere with future state.

We could also put in hooks for providers to do their own override logic?

File: provider/overrides/resources.rb [hunk 8]
```diff
@@ -19,6 +19,14 @@ module Overrides
     # This class allows them to get access to
     # Hash functions + lets the YAML parser import them.
     class OverrideResource < Google::YamlValidator
+      # This is the list of additional attributes that a provider
+      # can add through overrides
+      def self.attributes
+        []
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/overrides/resources.rb]: Yes it should. Because of how we marshal values in from YAML, initialize never gets called throughout our codebase. That's what our weird .validate functions do.

At some point, it would be nice to fix this, but that's hardly a priority.

File: provider/overrides/runner.rb [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+          end
+        end
+        prod.instance_variable_set('@objects',
+                                   old_prod.objects
+                                           .map { |o| build_resource(o, all_overrides[o.name]) })
+        prod
+      end
+
+      # Given a Api::Resource and Provider::Override::ResourceOverride,
+      # return a new Api::Resource with overrides applied.
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/overrides/runner.rb]: No, it'll change everything.

File: provider/overrides/runner.rb [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                 .each do |var_name|
+          if !res_override[var_name].nil?
+            res.instance_variable_set(var_name, res_override[var_name])
+          else
+            res.instance_variable_set(var_name, old_resource.instance_variable_get(var_name))
+          end
+        end
+
+        # Using instance_variable_get('properties') to make sure we get `exclude: true` properties
+        ['@properties', '@parameters'].each do |val|
```
[the remark this answers is not in the document]
Comment (chrisst) [reply provider/overrides/runner.rb]: I think previously we'd have to define `disk.encryption` as an override before we defined `disk.encryption.raw_key` otherwise it would try and insert a property into an object that didn't exist yet. If I'm following you correctly this means we will only have to define `disk.encryption.raw_key` and possibly remove some superfluous overrides?
Also I think I gave you a brief overview of how the disk fields are abusing overrides in order to work, but I imagine there will need to be some refactoring before they could use this system because they are super order dependent right now.

File: provider/overrides/runner.rb (hunk 3 above)
Comment (rambleraptor) [reply provider/overrides/runner.rb]: Static method it is.

File: provider/overrides/runner.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/overrides/runner.rb]: Yes, that should work.

**Comment by rambleraptor** [comment]
Updated!

**Comment by rambleraptor** [comment]
@SirGitsalot ptal
